ethereum/contracts/Scale.sol

```diff
@@ -0,0 +1,86 @@
+// SPDX-License-Identifier: MIT
+pragma solidity >=0.6.2;
+
+contract Scale {
+
+    // Read a byte at a specific index and return it as type uint8
+    function readByteAtIndex(bytes memory data, uint8 index)
+        public
+        pure
+        returns (uint8)
+    {
+        return uint8(data[index]);
+    }
+
+    // Reverse uint8[] array in place
+    function reverse(uint8[] memory arr)
+        public
+        pure
+        returns(uint8[] memory)
+    {
+        for (uint i = 0; i < arr.length/2; i++) {
+            uint8 current = arr[i];
+            uint256 otherIndex = arr.length - i - 1;
+            arr[i] = arr[otherIndex];
+            arr[otherIndex] = current;
+        }
+        return arr;
+    }
+
+    // Convert an uint8[] array to uint256
+    function uint8ArrayToUint256(uint8[] memory arr)
+        public
+        pure
+        returns (uint256)
+    {
+        uint256 number;
+        for(uint i = 0; i < arr.length; i++){
+            number = number+uint(arr[i])*(2**(8*(arr.length-(i+1))));
+        }
+        return number;
+    }
+
+    // Decodes a SCALE encoded compact unsigned integer
+    function decodeUintCompact(bytes memory data)
+        public
+        pure
+        returns (uint256)
+    {
+        uint8 b = readByteAtIndex(data, 0);           // read the first byte
+        uint8 mode = b & 3;                           // bitwise operation
+
+        if(mode == 0) {
+            return b >> 2;                           // right shift to remove mode bits
+        } else if(mode == 1) {
+            uint8 bb = readByteAtIndex(data, 1);      // read the second byte
+            uint64 r = bb;                            // convert to uint64
+            r <<= 6;                                  // multiply by * 2^6
+            r += b >> 2;                              // right shift to remove mode bits
+            return r;
+        } else if(mode == 2) {
+            uint8 b2 = readByteAtIndex(data, 1);      // read the next 3 bytes
+            uint8 b3 = readByteAtIndex(data, 2);
+            uint8 b4 = readByteAtIndex(data, 3);
+
+            uint32 x1 = uint32(b) | uint32(b2) << 8;  // convert to little endian
+            uint32 x2 = x1 | uint32(b3) << 16;
+            uint32 x3 = x2 | uint32(b4) << 24;
+
+            x3 >>= 2;                                 // remove the last 2 mode bits
+            return uint256(x3);
+        } else if(mode == 3) {
+            uint8 l = b >> 2;                         // remove mode bits
+            require(l <= 63,                          // Max upper bound of 536 is (67 - 4)
+                    "Not supported: l>63 encountered when decoding a compact-encoded uint");
```

Comment:
The biggest number u256 can hold is 32 bytes. I’d propose to just drop support for bigger numbers here and throw if l > 32

Reply:
That's definitely the easiest solution. Will dropping support for bigger numbers come back to bite us in the future?

Another potential option may be spreading the value across multiple memory slots i.e. 2+ uint256 variables.

Reply:
It will bite someone I guess, but every solution I can think of increases the gas cost for us, hence I would limit support for now.

The cleanest approach would likely be a big int type, which supports arithmetic operations, but that’s clearly out of scope.

Reply:
By rejecting all compact uints that take up greater than 32 bytes, all numbers that hit case 3 will be rejected. The smallest possible case 3 number is 1073741824 (decoded), which previously passed the test but is now rejected.

Should we just remove the case 3 logic?

Reply:
Is the logic working now as expected? If it is still buggy, I would remove it. If it works, let's keep it in there.

Reply:
Still buggy, removed in https://github.com/Snowfork/polkadot-ethereum/pull/51/commits/e9e7231e103fa771989a28187874db29300be165. We can bring it back and resolve the issues when applicable.